cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java

```diff
@@ -723,6 +723,8 @@
     // The following properties can be defined in cdap-site and overridden in individual pipeline config xml
     public static final String PIPELINE_BUFFER_SIZE = "log.process.pipeline.buffer.size";
     public static final String PIPELINE_EVENT_DELAY_MS = "log.process.pipeline.event.delay.ms";
+    public static final String PIPELINE_EVENT_OUT_OF_ORDER_MS = "log.process.pipeline.event.out.of.order.ms";
```

Comment:
In fact, we can stop the search if we see an event with `timestamp > (event timestamp to search for + event.delay.ms)`. And if we can't find the event we wanted, we don't need to resume from the beginning in Kafka, but rather from the largest offset that is smaller than the event timestamp (we can keep track of that during the linear scanning).